File: app/src/main/java/com/pddstudio/highlightjs/demo/utils/ThemeChangerDialog.java
package com.pddstudio.highlightjs.demo.utils;

import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;

import com.pddstudio.highlightjs.demo.R;
import com.pddstudio.highlightjs.models.Theme;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Created by pddstudio on 09/01/2017.
 */

public class ThemeChangerDialog implements DialogInterface.OnClickListener {

	public interface ThemeChangeListener {
		void onChangeTheme(@NonNull Theme theme);
	}

	private final ThemeChangeListener changeListener;
	private final Map<CharSequence, Theme> styleMap;
	private AlertDialog dialog;

	public ThemeChangerDialog(@NonNull ThemeChangeListener themeChangeListener) {
		this.changeListener = themeChangeListener;
		this.styleMap = new HashMap<>();
		initStyles();
	}

	private void initStyles() {
		for(Theme style : Theme.values()) {
			styleMap.put(style.getName(), style);
		}
	}

	public void show(@NonNull Context context) {
		if(dialog != null && dialog.isShowing()) {
			dialog.dismiss();
		}
		dialog = null;
		buildDialog(context);
		dialog.show();
	}

	private void buildDialog(@NonNull Context context) {
		dialog = new AlertDialog.Builder(context)
				.setSingleChoiceItems(buildThemeList(), -1, this)
				.setTitle(R.string.dialog_theme_selection_title)
				.create();
	}

	private CharSequence[] buildThemeList() {
		Set<CharSequence> themeNames = styleMap.keySet();
		return themeNames.toArray(new CharSequence[themeNames.size()]);
	}

	private Theme findThemeWithMatchingName(String themeName) {

		if(themeName == null || themeName.isEmpty()) {
			return Theme.DEFAULT;
		}

		for(Theme theme : Theme.values()) {
			if(theme.getName().equalsIgnoreCase(themeName)) {
				return theme;
			}
		}

		return Theme.DEFAULT;
	}

	private Theme getSelectedTheme(int selectedIndex) {

		CharSequence[] themes = buildThemeList();
		String themeName = null;

		if(selectedIndex > 0 && selectedIndex < themes.length) {
			themeName = themes[selectedIndex].toString();
		}

		return findThemeWithMatchingName(themeName);

	}

	@Override
	public void onClick(DialogInterface dialogInterface, int i) {
		dialogInterface.dismiss();
		Theme newTheme = getSelectedTheme(i);
		changeListener.onChangeTheme(newTheme);
	}
}

File: app/src/main/java/com/pddstudio/highlightjs/demo/utils/ThemeChangerDialog.kt
package com.pddstudio.highlightjs.demo.utils

import android.content.Context
import android.content.DialogInterface
import androidx.appcompat.app.AlertDialog
import com.pddstudio.highlightjs.demo.R
import com.pddstudio.highlightjs.models.Theme

/**
 * Created by pddstudio on 09/01/2017.
 */

class ThemeChangerDialog(private val changeListener: ThemeChangeListener) : DialogInterface.OnClickListener {

    interface ThemeChangeListener {
        fun onChangeTheme(theme: Theme)
    }

    private val styleMap: MutableMap<CharSequence, Theme> = HashMap()
    private var dialog: AlertDialog? = null

    init {
        initStyles()
    }

    private fun initStyles() {
        for (style in Theme.values()) {
            styleMap[style.getName()] = style
        }
    }

    fun show(context: Context) {
        if (dialog?.isShowing == true) {
            dialog?.dismiss()
        }
        dialog = null
        buildDialog(context)
        dialog?.show()
    }

    private fun buildDialog(context: Context) {
        dialog = AlertDialog.Builder(context)
            .setSingleChoiceItems(buildThemeList(), -1, this)
            .setTitle(R.string.dialog_theme_selection_title)
            .create()
    }

    private fun buildThemeList(): Array<CharSequence> {
        val themeNames = styleMap.keys
        return themeNames.toTypedArray()
    }

    private fun findThemeWithMatchingName(themeName: String?): Theme {
        if (themeName.isNullOrEmpty()) return Theme.DEFAULT
        for (theme in Theme.values()) {
            if (theme.getName().equals(themeName, ignoreCase = true)) return theme
        }
        return Theme.DEFAULT
    }

    private fun getSelectedTheme(selectedIndex: Int): Theme {
        val themes = buildThemeList()
        val themeName = if (selectedIndex > 0 && selectedIndex < themes.size) {
            themes[selectedIndex].toString()
        } else null
        return findThemeWithMatchingName(themeName)
    }

    override fun onClick(dialogInterface: DialogInterface, i: Int) {
        dialogInterface.dismiss()
        val newTheme = getSelectedTheme(i)
        changeListener.onChangeTheme(newTheme)
    }
}
